src/providers/arcgisrest/qgsafsprovider.cpp

```diff
@@ -317,9 +322,240 @@ QgsLayerMetadata QgsAfsProvider::layerMetadata() const
+  QgsReadWriteLocker locker( mSharedData->mReadWriteLock, QgsReadWriteLocker::Read );
+  for ( auto it = geometryMap.constBegin(); it != geometryMap.constEnd(); ++it )
+  {
+    const QgsFeatureId id = it.key();
+    QgsFeature feature( fields );
+    feature.setId( id );
+    // we ONLY require the objectId field set here
+    feature.setAttribute( objectIdFieldIndex, mSharedData->featureIdToObjectId( id ) );
+    feature.setGeometry( it.value() );
+
+    updatedFeatures.append( feature );
+  }
+  locker.unlock();
+
+  QString error;
+  QgsFeedback feedback;
+  const bool res = mSharedData->updateFeatures( updatedFeatures, true, false, error, &feedback );
+  if ( !res )
+    pushError( tr( "Error while updating features: %1" ).arg( error ) );
+
+  return res;
+}
+
+bool QgsAfsProvider::changeFeatures( const QgsChangedAttributesMap &attrMap, const QgsGeometryMap &geometryMap )
+{
+  if ( !mCapabilityStrings.contains( QLatin1String( "update" ), Qt::CaseInsensitive ) )
+    return false;
+
+  QgsFeatureIds ids;
+  ids.reserve( attrMap.size() + geometryMap.size() );
+  for ( auto it = attrMap.constBegin(); it != attrMap.constEnd(); ++it )
+  {
+    const QgsFeatureId id = it.key();
+    ids << id;
+  }
+  for ( auto it = geometryMap.constBegin(); it != geometryMap.constEnd(); ++it )
+  {
+    const QgsFeatureId id = it.key();
+    ids << id;
+  }
+
+  // REST API requires a full definition of features, so we have to read their initial values first
+  QgsFeatureIterator it = getFeatures( QgsFeatureRequest().setFilterFids( ids ) );
+  QgsFeature feature;
+
+  QgsFeatureList updatedFeatures;
+  updatedFeatures.reserve( attrMap.size() );
+  const int objectIdFieldIndex = mSharedData->mObjectIdFieldIdx;
+
+  while ( it.nextFeature( feature ) )
+  {
+    QgsFeature modifiedFeature = feature;
+
+    const QgsAttributeMap modifiedAttributes = attrMap.value( feature.id() );
+    for ( auto aIt = modifiedAttributes.constBegin(); aIt != modifiedAttributes.constEnd(); ++aIt )
+    {
+      if ( aIt.key() == objectIdFieldIndex )
+        continue; // can't modify the objectId field!
+
+      modifiedFeature.setAttribute( aIt.key(), aIt.value() );
+    }
+
+    const auto geomIt = geometryMap.constFind( feature.id() );
+    if ( geomIt != geometryMap.constEnd() )
+    {
+      modifiedFeature.setGeometry( geomIt.value() );
+    }
+
+    updatedFeatures.append( modifiedFeature );
+  }
+
+  QString error;
+  QgsFeedback feedback;
+  const bool res = mSharedData->updateFeatures( updatedFeatures, true, true, error, &feedback );
+  if ( !res )
+    pushError( tr( "Error while updating features: %1" ).arg( error ) );
+
+  return res;
+}
+
+bool QgsAfsProvider::addAttributes( const QList<QgsField> &attributes )
+{
+  if ( mAdminUrl.isEmpty() )
+    return false;
+
+  const QStringList adminCapabilities = mAdminData.value( QStringLiteral( "capabilities" ) ).toString().split( ',' );
+  if ( !adminCapabilities.contains( QLatin1String( "update" ), Qt::CaseInsensitive ) )
```

Comment:
Just for my personal education, why did esri have both capabilities vs admin capabilities covering what appears to be the same thing (feature addition / editing / deletion vs. attributes addition / deletion?)

Also, would it be worth coming up with a mAdminCapabilityStrings the same way you did mCabilityStrings?

Reply:
> Just for my personal education, why did esri have both capabilities vs admin capabilities covering what appears to be the same thing 

They are different API endpoints -- one is for layer edits, the other for table maintenance. I suspect many users only get access to the first and can manage features but not edit the structure of tables.